### PR TITLE
Token Testing Feature Changes

### PR DESCRIPTION
**Pull request by pkusumaraju-rythmos**
Following changes will generate token and parent token results for the input files.
Also compares with respective output file, if files are identical test case will be passes otherwise test case will be failed.

Cmake changes will ensure to execute the specified language test suit which we need to run on Uncrustify for tokens or parent tokens or simple output files.

### CODE REVIEW

**Comment by unitydevtools** [comment]
Can one of the admins verify this patch?

File: .gitignore [hunk 1]
```diff
@@ -99,3 +101,4 @@ build/
 .cproject
 .project
 .settings
+src/uncrustify.cpp
```
Comment (mihaipopescu) [inline .gitignore]: looks like a mistake maybe?

File: .gitignore [hunk 2]
```diff
@@ -7,6 +7,8 @@ src/uncrustify_version.h
 results/
 tests/results/
 tests/results_2/
+tests/ptok_results/
+tests/tok_results/
```
Comment (mihaipopescu) [inline .gitignore]: unclear what ptok/tok are here. Also `results_2` is suspicious bt out of scope

File: src/uncrustify.cpp [hunk 3]
```diff
@@ -104,7 +104,7 @@ static bool ends_with(const char *filename, const char *tag, bool case_sensitive
  * @param no_backup    don't create a backup, if filename_out == filename_in
  * @param keep_mtime   don't change the mtime (dangerous)
  */
-static void do_source_file(const char *filename_in, const char *filename_out, const char *parsed_file, bool no_backup, bool keep_mtime);
+static void do_source_file(const char *filename_in, const char *filename_out, const char *parsed_file, const char *token_output_file, const char *ptoken_output_file, bool no_backup, bool keep_mtime);
```
Comment (mihaipopescu) [inline src/uncrustify.cpp:107]: I'd just use the previous `parsed_file` but use a new export format or something like this (more on this later).

**Review by mihaipopescu — CHANGES_REQUESTED** [review]
There are a couple of things I don't quite understand wrt the design of this feature. While I understand that maybe we could probably use another format in storing the "AST" token dump which is more friendly to tools than to humans, I would not try to lose the previous information nor split them in multiple files (like now there seems like we're going to have another file for parents).

I would probably choose something like json or even much simpler csv (simpler to do the parser) and use that extension to avoid confusion in editors. Like that would be simpler to choose what to check and what not from the test code (say that we initially we care only about token type and parent token and don't care about flags or we could add more checks as things evolve).

Also, I would give a very good look at our existing cli tests https://github.com/Unity-Technologies/uncrustify/blob/master/tests/cli/test_cli_options.py as they are already checking full token dump (`-p`) here output https://github.com/Unity-Technologies/uncrustify/blob/master/tests/cli/Output/p.txt so I'm actually unsure if we should even build a  thing or extend the existing fw.

**Comment by mihaipopescu** [comment]
test this please

**Comment by pkusumaraju-rythmos** [comment]
test this please

**Comment by hschlichter** [comment]
Where are we with this? @mihaipopescu have your comments been addressed? And if not, what is needed so we can move this forward?

File: Token_help.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
+## Token Generation
+- In cmake test suit we have integrated a new feature to generate token and parent token and the compare the output and result file. If both the files are same then test case will pass otherwise testcase will be failed.
+- Generating Token files 
+- we can generate Token files by using the  command line option ```--token``` followed by token file name
+	- Eg: ``` uncrustify.exe -c Uncrustify.Common-CStyle.cfg -p debug.txt -f input.cpp -o output.cpp --token Tokens.txt -La  > a.log 2>&1```
```
Comment (mihaipopescu) [inline Token_help.md:5]: single back quote or multiline triple backquotes here

File: Token_help.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+			3    [FUNC_CALL]         )
+			3    [FUNC_CALL]         {
+			3    [NONE]              
+			4    [FUNC_CALL]         }
+		```
+
+- we can generate generate Tokens & Parent Tokens on one go by enabling both --token,--parent
+	- Eg: ``` uncrustify.exe -c Uncrustify.Common-CStyle.cfg -p debug.txt -f ip.cpp -o op.txt --token Tokens.txt --parent parent_token.txt -La  > a.log 2>&1```
+
+### Cmake Changes for RUN_TESTS
```
Comment (mihaipopescu) [inline Token_help.md:43]: I have no idea what this section is about and why it's needed

**Review by mihaipopescu — CHANGES_REQUESTED** [review]

File: Token_help.md (hunk 5 above)
Comment (spuppala-rythmos) [reply Token_help.md:43]: This is just an example for generating tokens and parent tokens both.

Comment (mihaipopescu) [reply Token_help.md:43]: No, I meant below

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -11,20 +11,38 @@ function(add_uncrustify_test name lang config rerun_config input result result_2
       -DTEST_RESULT=${result}
       -DTEST_RESULT_2=${result_2}
       -DTEST_OUTPUT=${output}
+      -DTEST_TYPE=${test_do}
```
Comment (mihaipopescu) [inline tests/CMakeLists.txt]: what does _do_ from `test_do` mean here?

File: tests/CMakeLists.txt [hunk 7]
```diff
@@ -11,20 +11,38 @@ function(add_uncrustify_test name lang config rerun_config input result result_2
...
       -DTEST_RERUN_OUTPUT=${rerun_output}
       -DTEST_DIR=${PROJECT_SOURCE_DIR}/tests
-      -P ${PROJECT_SOURCE_DIR}/tests/run_test.cmake
+      -P ${PROJECT_SOURCE_DIR}/tests/run_test_all.cmake
```
Comment (mihaipopescu) [inline tests/CMakeLists.txt]: what will happen with the old script?

**Review by mihaipopescu — CHANGES_REQUESTED** [review]
Let me reiterate my questions:

1. Why create another dump function where we can use the existing one (+ tuning/filter params if necessary)?
2. Why create a new framework when we already have a cli framework that does this already? (see `tests/cli/test_cli_options.py` and `tests/cli/Output/p.txt`).
3. Why create two dump files one for parent type and one for token type?
4. Why use the original file extension for the dump files? An extension like `.txt` would not confuse the editors (and humans).
5. Why abandon and not delete the old test cmake script? It will become obsolete.

Testing:
* How does the change impacts the speed of running the test framework in terms of the performance?
* How much maintenance is required to update the dump test files in current development?

File: tests/CMakeLists.txt (hunk 7 above)
Comment (spuppala-rythmos) [reply tests/CMakeLists.txt]: we can delete the old script ... just for bkup kept that file.

File: tests/CMakeLists.txt (hunk 6 above)
Comment (spuppala-rythmos) [reply tests/CMakeLists.txt]: this one needs a meaning full name ... will change and upload new file

**Pull request closed without merging**